File: 0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == nullptr || head->next == nullptr) return false;
        
        ListNode* slowPtr = head; // Acts also as current
        ListNode* fastPtr = head->next;
        
        while(true){
            if(slowPtr == nullptr || fastPtr == nullptr || fastPtr->next == nullptr) return false;
            if(slowPtr == fastPtr) return true;
            slowPtr = slowPtr->next;
            fastPtr = fastPtr->next->next;
        }
    }
};

File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None or head.next is None:
            return False
        
        slow, fast = head, head.next # slow also serves as current
        
        while True:
            if fast is None or fast.next is None:
                return False
            if slow == fast:
                return True
            
            # fast traverses twice as fast
            slow = slow.next
            fast = fast.next.next